### DISCUSSION TITLE
95. 인덱스는 insert할 때와 delete할 때 왜 성능이 저하되나요?

### DISCUSSION THREAD

**Discussion by jongmee** [jewelry-box/cs-interview#99] [General]
인덱스는 insert할 때와 delete할 때 왜 성능이 저하되나요?

**Comment by linirini** [comment]
### insert
레코드가 추가되었을 때 항상 정렬된 상태를 유지하기 위해서 추가될 인덱스의 적절한 위치를 탐색한 후에 저장한다. 만약 레코드를 삽입할 때 인덱스 페이지의 공간이 부족하면, 페이지가 두 개로 나뉘어야 하는 "페이지 분할"이 발생한다. 이는 추가적인 I/O를 발생시키고, 성능 저하를 초래할 수 있다.
InnoDB에서는 메모리에 인덱스 추가 작업을 지연시켜서 디스크 쓰기 횟수를 줄여 효율적으로 한 번에 쓰도록 할 수 있다.
>1. 지연된 쓰기 (Delayed Write):
버퍼 풀: InnoDB는 '버퍼 풀'이라는 메모리 영역을 사용하여 데이터 페이지와 인덱스 페이지를 저장합니다. 트랜잭션이 발생하면 이러한 변경 내용이 먼저 버퍼 풀에 저장되고, 이후 배치 처리로 디스크에 쓰여집니다. 이는 디스크 I/O 작업을 최소화하여 성능을 향상시킵니다.
>2. 배치 쓰기 처리:
한 번에 쓰기: 여러 변경 사항이 메모리에 저장된 후, 디스크에 한 번에 더 많은 데이터를 쓰게 됩니다. 이로 인해 디스크 I/O 작업 횟수가 줄어들고 성능이 개선됩니다. 특히 데이터베이스에 대한 동시 요청이 많은 경우, 이러한 배치 처리는 효과적입니다.

이러한, 지연된 쓰기와 배치 처리를 통해 I/O 병목 현상을 줄이고, 결과적으로 데이터베이스의 전반적인 성능이 향상된다.

### delete
레코드 삭제 시에도 인덱스가 삭제되어야 한다. 이때, 인덱스 삭제 방법은 해당 레코드가 저장된 인덱스의 리프 노드에 마킹하여 수행한다. 따라서, 삭제 마킹된 인덱스 키 공간은 그대로 방치하거나, 재활용할 수 있다. 대신, 메모리 공간을 차지하고 있으므로, 메모리 자원을 효율적으로 사용하지 못할 수 있다.
